flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java

```diff
@@ -0,0 +1,228 @@
+  private static final Logger LOG = LoggerFactory.getLogger(DataFileRewritePlanner.class);
+
+  private final String tableName;
+  private final String taskName;
+  private final int taskIndex;
+  private final TableLoader tableLoader;
+  private final int partialProgressMaxCommits;
+  private final long maxRewriteBytes;
+  private final Map<String, String> rewriterOptions;
+  private transient SizeBasedDataRewriter rewriter;
+  private transient RewriteFileGroupPlanner planner;
+  private transient Counter errorCounter;
+
+  public DataFileRewritePlanner(
+      String tableName,
+      String taskName,
+      int taskIndex,
+      TableLoader tableLoader,
+      int newPartialProgressMaxCommits,
+      long maxRewriteBytes,
+      Map<String, String> rewriterOptions) {
+    Preconditions.checkNotNull(tableName, "Table name should no be null");
+    Preconditions.checkNotNull(taskName, "Task name should no be null");
+    Preconditions.checkNotNull(tableLoader, "Table loader should no be null");
+    Preconditions.checkNotNull(rewriterOptions, "Options map should no be null");
+
+    this.tableName = tableName;
+    this.taskName = taskName;
+    this.taskIndex = taskIndex;
+    this.tableLoader = tableLoader;
+    this.partialProgressMaxCommits = newPartialProgressMaxCommits;
+    this.maxRewriteBytes = maxRewriteBytes;
+    this.rewriterOptions = rewriterOptions;
+  }
+
+  @Override
+  public void open(Configuration parameters) throws Exception {
+    tableLoader.open();
+    this.errorCounter =
+        TableMaintenanceMetrics.groupFor(getRuntimeContext(), tableName, taskName, taskIndex)
+            .counter(TableMaintenanceMetrics.ERROR_COUNTER);
+
+    this.rewriter =
+        new SizeBasedDataRewriter(tableLoader.loadTable()) {
+          @Override
+          public Set<DataFile> rewrite(List<FileScanTask> group) {
+            // We use the rewriter only for bin-packing the file groups to compact
+            throw new UnsupportedOperationException("Should not be called");
+          }
+        };
+
+    rewriter.init(rewriterOptions);
+    this.planner = new RewriteFileGroupPlanner(rewriter, RewriteJobOrder.NONE);
+  }
+
+  @Override
+  public void processElement(Trigger value, Context ctx, Collector<PlannedGroup> out)
+      throws Exception {
+    LOG.debug(
+        "Creating rewrite plan for table {} with {}[{}] at {}",
+        tableName,
+        taskName,
+        taskIndex,
+        ctx.timestamp());
+    try {
+      SerializableTable table =
+          (SerializableTable) SerializableTable.copyOf(tableLoader.loadTable());
+      if (table.currentSnapshot() == null) {
+        LOG.info(
+            "Nothing to plan for in an empty table {} with {}[{}] at {}",
+            tableName,
+            taskName,
+            taskIndex,
+            ctx.timestamp());
+        return;
+      }
+
+      RewritePlanResult plan =
+          planner.plan(table, Expressions.alwaysTrue(), table.currentSnapshot().snapshotId());
```

Comment:
`startingSnapshotId` is the snapshot when the compaction started (files from newer snapshots are ignored during this run). This is the snapshotId which will be used for the new files generated by the compaction.
The name is coming from the existing code, and I tried to avoid modifying core/Spark code as much as possible